### DISCUSSION TITLE
Use MIDI to make a drum pattern with drum samples

### DISCUSSION THREAD

**Discussion by samuelbraun04** [DBraun/DawDreamer#137] [Q&A] [answered]
I have a hi hat MIDI file and a hi hat sample file. I was wondering how I could make a clip where the hi hat sample is played to the midi notes? I've tried:

```
hihat = engine.make_playback_processor("hihat", self.load_audio_file(pathHihat))
hihat.load_midi(midiHihat, beats=True)
```

but that obviously doesn't work, but the idea is sort of there. Thank you.

**Comment by DBraun** [comment, marked as answer]
One way is to use the [`SamplerProcessor`](https://dirt.design/DawDreamer/_generate/dawdreamer.SamplerProcessor.html), which is missing good examples...

```python
sampler = engine.make_sampler_processor("sampler", self.load_audio_file(pathHihat))
# later you can call set_data
# sampler.set_data(self.load_audio_file(pathHihat))
sampler.load_midi(midiHihat, beats=True)
```

However, that sampler will pitch the notes according to the MIDI note, which you may not want. It also comes with many ADSR features which you'll have to set to your needs.

I'd also recommend adapting the polyphonic Faust example here: https://github.com/DBraun/DawDreamer/blob/main/tests/test_faust_poly_sampler.py
and Faust code here: https://github.com/DBraun/DawDreamer/blob/main/tests/faust_dsp/polyphonic_sampler.dsp

It currently plays back a cymbal with polyphony, and the MIDI note affects the pitch and speed of the sample. To not let the MIDI not affect the pitch/speed, you'd change the definition of ratio to `ratio = 1;` Then the MIDI note of 60 would be a neutral playback speed.

You may want to remove `envVol`. It's just there as an idea for applying an amplitude envelope to each note.

**Reply by samuelbraun04** [reply]
The SamplerProcessor was exactly what I was looking for – thank you. However, when I do the following:

```
        hihat = self.engine.make_sampler_processor("hihat", self.convertAudioFile(self.pathHihat))
        hihat.set_data(self.convertAudioFile(self.pathHihat))
        hihat.load_midi(self.midiHihat, beats=True)

        hihat_graph = [
            (hihat, [])
        ]

        self.engine.load_graph(hihat_graph)
        self.engine.render(16., beats=True)
        wavfile.write(self.WORKSPACE_CONTAINER_PATH+'\\'+'hihat.wav', self.SAMPLE_RATE, (self.engine.get_audio()).transpose())
```

I get an audio file that plays short digital clicks in the pattern of the midi file. Is there something I'm missing?

**Reply by DBraun** [reply]
Maybe you need to transpose the result from convertAudioFile too.

**Reply by samuelbraun04** [reply]
Yes that seems to be the problem – how can I tranpose the MIDI down? If I transpose the exported file it looks a lot of the quality.

Also, how can I get rid of the clicking that occurs between the instances of the sample playing?  (The audio: http://sndup.net/bzqs)

**Reply by samuelbraun04** [reply]
Found a way to transpose the MIDI keys via the [music21](https://web.mit.edu/music21/doc/) module.

However, I'm still getting the clicks between sounds when they play right after one another (as seen in the previous reply's example).

**Reply by samuelbraun04** [reply]
Figured everything out!
